### PR TITLE
[Feature] record memory cost in audit log

### LINKED ISSUE #6352: Record memory cost in audit log
## Feature request

**Is your feature request related to a problem? Please describe.**

The execution cost of a query should be recorded into the audit log, for further analysis.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by murphyatwork**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #6352

## Problem Summary(Required) ：

Record memory costs in audit log for further analysis.

## TODO
Current memory computation just sum up the peak consumption of all query mem-tracker, in some cases it could be imprecise, without considering the shared memory between some fragments.

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 8 / 12 (66.67%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/plugin/AuditEvent.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/qe/ConnectProcessor.java|2|2|100.00%|

File: be/src/runtime/query_statistics.cpp [hunk 1]
```diff
@@ -23,10 +23,28 @@
...
+}
+
 void QueryStatistics::merge(QueryStatisticsRecvr* recvr) {
     recvr->merge(this);
 }
 
+void QueryStatistics::merge_pb(const PQueryStatistics& statistics) {
+    scan_rows += statistics.scan_rows();
+    scan_bytes += statistics.scan_bytes();
+    cpu_ns += statistics.cpu_cost_ns();
```
Comment (trueeyu) [inline be/src/runtime/query_statistics.cpp:43]: add or max ?

File: be/src/exec/vectorized/hash_join_node.cpp [hunk 2]
```diff
@@ -56,11 +56,17 @@ Status HashJoinNode::init(const TPlanNode& tnode, RuntimeState* state) {
 
     const std::vector<TEqJoinCondition>& eq_join_conjuncts = tnode.hash_join_node.eq_join_conjuncts;
     for (const auto& eq_join_conjunct : eq_join_conjuncts) {
-        ExprContext* ctx = nullptr;
-        RETURN_IF_ERROR(Expr::create_expr_tree(_pool, eq_join_conjunct.left, &ctx));
-        _probe_expr_ctxs.push_back(ctx);
-        RETURN_IF_ERROR(Expr::create_expr_tree(_pool, eq_join_conjunct.right, &ctx));
-        _build_expr_ctxs.push_back(ctx);
+        ExprContext* left = nullptr;
```
Comment (trueeyu) [inline be/src/exec/vectorized/hash_join_node.cpp:59]: split to another pr?

Comment (mofeiatwork) [reply be/src/exec/vectorized/hash_join_node.cpp:59]: That's a dependency on the other PR. I would rebase after it merged.

File: be/src/runtime/query_statistics.cpp (hunk 1 above)
Comment (mofeiatwork) [reply be/src/runtime/query_statistics.cpp:43]: It's used to merge statistic from other query fragments, so it should be `add` instead of max.

File: be/src/exec/pipeline/result_sink_operator.cpp [hunk 3]
```diff
@@ -49,6 +49,8 @@ void ResultSinkOperator::close(RuntimeState* state) {
             auto query_statistic = std::make_shared<QueryStatistics>();
             QueryContext* query_ctx = state->query_ctx();
             query_statistic->add_scan_stats(query_ctx->cur_scan_rows_num(), query_ctx->get_scan_bytes());
+            query_statistic->add_cpu_costs(query_ctx->cpu_cost());
```
Comment (trueeyu) [inline be/src/exec/pipeline/result_sink_operator.cpp:52]: Cpu time is sum of all the runtime time of all fragments ?

Comment (trueeyu) [inline be/src/exec/pipeline/result_sink_operator.cpp:52]: Cpu time is sum of all the runtime time of all fragments ?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 4 / 6 (66.67%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/plugin/AuditEvent.java|3|3|100.00%|
|:large_blue_circle:|com/starrocks/qe/ConnectProcessor.java|1|1|100.00%|

**Comment by mofeiatwork** [comment]
@Mergifyio rebase

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 4 / 14 (28.57%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/common/util/DebugUtil.java|0|3|00.00%|
|:large_blue_circle:|com/starrocks/plugin/AuditEvent.java|3|3|100.00%|
|:large_blue_circle:|com/starrocks/qe/ConnectProcessor.java|1|1|100.00%|

File: be/src/runtime/query_statistics.cpp (hunk 1 above)
Comment (stdpain) [reply be/src/runtime/query_statistics.cpp:43]: if a plan is 
```
fragment 1
JOIN
|-EXCHANGE(id=1)
EXCHANGE 

fragment 2
SINK(dst=1)
|
JOIN
|-EXCHANGE(id=2)
|
EXCHANGE

fragment 3
SINK(dst=2)
|
AGG
```

real mem peak cost is max((fragment 1 + fragment 2), (fragment2, fragment3))

But I don't have a better idea

Comment (stdpain) [reply be/src/runtime/query_statistics.cpp:43]: In a complex plan, this algorithm has the potential to give a greater memory consumption statistic.

**Review by kangkaisen — APPROVED** [review]

File: be/src/runtime/query_statistics.cpp (hunk 1 above)
Comment (mofeiatwork) [reply be/src/runtime/query_statistics.cpp:43]: Add TODO in description, and leave it for future work.

**Review by stdpain — APPROVED** [review]
